toolchain/check/convert.cpp

```diff
@@ -1088,8 +1088,18 @@ auto ConvertCallArgs(Context& context, Parse::Node call_parse_node,
 
 auto ExpressionAsType(Context& context, Parse::Node parse_node,
                       SemIR::InstId value_id) -> SemIR::TypeId {
-  return context.CanonicalizeType(ConvertToValueOfType(
-      context, parse_node, value_id, SemIR::TypeId::TypeType));
+  auto type_inst_id = ConvertToValueOfType(context, parse_node, value_id,
+                                           SemIR::TypeId::TypeType);
+  if (type_inst_id == SemIR::InstId::BuiltinError) {
+    return SemIR::TypeId::Error;
+  }
+  auto type_id = context.CanonicalizeType(type_inst_id);
+  if (type_id == SemIR::TypeId::Error) {
+    CARBON_DIAGNOSTIC(TypeConversionFailure, Error,
+                      "Cannot convert expression to type `type`.");
```

Comment:
Would something like "Cannot evaluate type expression" work better here? We only get here if the actual conversion to type `type` succeeded.

Reply:
Done